### PR TITLE
[Doc] Add liquid clustering preview documentation

### PR DESCRIPTION
**Pull request by zedtang**
#### Which Delta project/connector is this regarding?

- [ ] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [x] Other (Doc)

## Description

This PR adds documentation about liquid clustering preview in Delta Lake 3.1.0.

## How was this patch tested?

127 0 0 1_8000_delta-clustering html (2)

### CODE REVIEW

**Review by vkorukanti — COMMENTED** [review]
We have an optimizations page [here](https://docs.delta.io/latest/optimizations-oss.html#z-ordering-multi-dimensional-clustering). Could you please add a section briefly mentioning liquid and a pointer to this page?

File: docs/source/delta-clustering.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
+---
+description: Learn about liquid clustering in <Delta>.
+orphan: 1
+---
+
+# Use liquid clustering for Delta tables
+
+liquid clustering replaces table partitioning and `ZORDER` to simplify data layout decisions and optimize query performance. Liquid clustering provides flexibility to redefine clustering keys without rewriting existing data, allowing data layout to evolve alongside analytic needs over time.
```
Comment (imback82) [inline docs/source/delta-clustering.md:8]:
```suggestion
Liquid clustering improves the existing partitioning and `ZORDER` techniques by simplifying data layout decisions in order to optimize query performance. Liquid clustering provides flexibility to redefine clustering keys without rewriting existing data, allowing data layout to evolve alongside analytic needs over time.
```

File: docs/source/delta-clustering.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+```sql
+-- Create an empty table
+CREATE TABLE table1(col0 int, col1 string) USING DELTA CLUSTER BY (col0);
+
+-- Using a CTAS statement
+CREATE TABLE table2 CLUSTER BY (col0)  -- specify clustering after table name, not in subquery
+AS SELECT * FROM table1;
+```
+
+.. warning:: Tables created with liquid clustering enabled have Clustering table feature enabled at creation and use Delta writer version 7 and reader version 3. Table protocol versions cannot be downgraded, and tables with clustering enabled are not readable by <Delta> clients that do not support all enabled Delta reader protocol table features. See [_](/versioning.md).
```
Comment (imback82) [inline docs/source/delta-clustering.md:37]: we don't have any reader-writer feature enabled for clustering table feature?

File: docs/source/delta-clustering.md [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+| Current data optimization technique | Recommendation for clustering keys |
+| --- | --- |
+| Hive-style partitioning | Use partition columns as clustering keys. |
+| Z-order indexing | Use the `ZORDER BY` columns as clustering keys. |
+| Hive-style partitioning and Z-order | Use both partition columns and `ZORDER BY` columns as clustering keys. |
+| Generated columns to reduce cardinality (for example, date for a timestamp) | Use the original column as a clustering key, and don't create a generated column. |
+
+## Write data to a clustered table
+
+You must use a Delta writer client that supports Clustering table feature.
```
Comment (imback82) [inline docs/source/delta-clustering.md:54]: and domain metadata?

File: docs/source/delta-clustering.md [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+## <a id="optimize"></a> How to trigger clustering
+
+Use the `OPTIMIZE` command on your table, as in the following example:
+
+```sql
+OPTIMIZE table_name;
+```
+
+Liquid clustering is incremental, meaning that data is only rewritten as necessary to accommodate data that needs to be clustered. Data files with clustering keys that do not match data to be clustered are not rewritten.
```
Comment (imback82) [inline docs/source/delta-clustering.md:64]: this is not available yet?

File: docs/source/delta-clustering.md [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+## Read data from a clustered table
+
+You can read data in a clustered table using any <Delta> client. For best query results, include clustering keys in your query filters, as in the following example:
```
Comment (imback82) [inline docs/source/delta-clustering.md:68]: this contradicts the statement above where I left a comment.

File: docs/source/delta-clustering.md [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+## Change clustering keys
+
+You can change clustering keys for a table at any time by running an `ALTER TABLE` command, as in the following example:
+
+```sql
+ALTER TABLE table_name CLUSTER BY (new_column1, new_column2);
+```
+
+When you change clustering keys, subsequent `OPTIMIZE` and write operations use the new clustering approach, but existing data is not rewritten.
```
Comment (imback82) [inline docs/source/delta-clustering.md:82]: this (existing data is not re-written) is not true yet?

File: docs/source/delta-clustering.md [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+```sql
+ALTER TABLE table_name CLUSTER BY NONE;
+```
+
+Setting cluster keys to `NONE` does not rewrite data that has already been clustered, but prevents future `OPTIMIZE` operations from using clustering keys.
+
+## See how table is clustered
+
+You can use `DESCRIBE DETAIL` commands to see the clustering keys for a table, as in the following examples:
```
Comment (imback82) [inline docs/source/delta-clustering.md:94]: do we support this already?

File: docs/source/delta-clustering.md [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+```sql
+SELECT * FROM table_name WHERE cluster_key_column_name = "some_value";
+```
+
+## Change clustering keys
+
+You can change clustering keys for a table at any time by running an `ALTER TABLE` command, as in the following example:
+
+```sql
+ALTER TABLE table_name CLUSTER BY (new_column1, new_column2);
```
Comment (imback82) [inline docs/source/delta-clustering.md:79]: not yet supported?

**Comment by zedtang** [comment]
> We have an optimizations page [here](https://docs.delta.io/latest/optimizations-oss.html#z-ordering-multi-dimensional-clustering). Could you please add a section briefly mentioning liquid and a pointer to this page?

Might be better to do this for 3.2 since incremental clustering is not supported yet.

File: docs/source/delta-clustering.md [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+- Tables often filtered by high cardinality columns.
+- Tables with significant skew in data distribution.
+- Tables that grow quickly and require maintenance and tuning effort.
+- Tables with access patterns that change over time.
+- Tables where a typical partition key could leave the table with too many or too few partitions.
+
+## Enable liquid clustering
+
+You must enable liquid clustering when first creating a table. Clustering is not compatible with partitioning or `ZORDER`. Once enabled, run `OPTIMIZE` jobs as normal to cluster data. See [_](#optimize).
```
Comment (imback82) [inline docs/source/delta-clustering.md:24]:
```suggestion
You must enable liquid clustering when creating a table. Clustering is not compatible with partitioning or `ZORDER`. Once enabled, run `OPTIMIZE` jobs as normal to cluster data. See [_](#optimize).
```

File: docs/source/delta-clustering.md [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+```sql
+-- Create an empty table
+CREATE TABLE table1(col0 int, col1 string) USING DELTA CLUSTER BY (col0);
+
+-- Using a CTAS statement
+CREATE TABLE table2 CLUSTER BY (col0)  -- specify clustering after table name, not in subquery
+AS SELECT * FROM table1;
+```
+
+.. warning:: Tables created with liquid clustering enabled have Clustering and Domain metadata table feature enabled at creation and use Delta writer version 7 and reader version 3. Table protocol versions cannot be downgraded. See [_](/versioning.md).
```
Comment (imback82) [inline docs/source/delta-clustering.md:37]:
```suggestion
.. warning:: Tables created with liquid clustering have `Clustering` and `DomainMetadata` table features enabled (both writer features) and use Delta writer version 7 and reader version 3. Table protocol versions cannot be downgraded. See [_](/versioning.md).
```

File: docs/source/delta-clustering.md [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+| Current data optimization technique | Recommendation for clustering keys |
+| --- | --- |
+| Hive-style partitioning | Use partition columns as clustering keys. |
+| Z-order indexing | Use the `ZORDER BY` columns as clustering keys. |
+| Hive-style partitioning and Z-order | Use both partition columns and `ZORDER BY` columns as clustering keys. |
+| Generated columns to reduce cardinality (for example, date for a timestamp) | Use the original column as a clustering key, and don't create a generated column. |
+
+## Write data to a clustered table
+
+You must use a Delta writer client that supports Clustering and Domain metadata table feature.
```
Comment (imback82) [inline docs/source/delta-clustering.md:54]:
```suggestion
You must use a Delta writer client that supports `Clustering` and `DomainMetadata` table features.
```

File: docs/source/delta-clustering.md (hunk 5 above)
Comment (imback82) [reply docs/source/delta-clustering.md:69]: need to specify minimum protocol version? (not sure how other examples address this)

File: docs/source/delta-clustering.md [hunk 12]
```diff
@@ -0,0 +1,113 @@
+---
+description: Learn about liquid clustering in <Delta>.
+orphan: 1
+---
+
+# Use liquid clustering for Delta tables
+
+Liquid clustering improves the existing partitioning and `ZORDER` techniques by simplifying data layout decisions in order to optimize query performance. Liquid clustering provides flexibility to redefine clustering keys without rewriting existing data, allowing data layout to evolve alongside analytic needs over time.
```
Comment (imback82) [inline docs/source/delta-clustering.md:8]: can we change "key" -> "column" in all instances?

File: docs/source/delta-clustering.md [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+``` -->
+
+## Limitations
+
+The following limitations exist:
+
+- In <Delta> 3.1, users needs to enable the feature flag `spark.databricks.delta.clusteredTable.enableClusteringTablePreview` to use liquid clustering. The following features are not supported in this preview:
+  - Incremental clustering
+  - ALTER TABLE ... CLUSTER BY
+  - DESCRIBE DETAIL
```
Comment (imback82) [inline docs/source/delta-clustering.md:109]:
```suggestion
  - Z-cube based incremental clustering
  - `ALTER TABLE ... CLUSTER BY` to change clustering columns
  - `DESCRIBE DETAIL` to inspect the current clustering columns
```

File: docs/source/delta-clustering.md (hunk 5 above)
Comment (zedtang) [reply docs/source/delta-clustering.md]: Added reader version requirements

File: docs/source/delta-clustering.md [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+
+## Limitations
+
+The following limitations exist:
+
+- In <Delta> 3.1, users needs to enable the feature flag `spark.databricks.delta.clusteredTable.enableClusteringTablePreview` to use liquid clustering. The following features are not supported in this preview:
+  - ZCube based incremental clustering
+  - `ALTER TABLE ... CLUSTER BY` to change clustering columns
+  - `DESCRIBE DETAIL` to inspect the current clustering columns
+- You can only specify columns with statistics collected for clustering keys. By default, the first 32 columns in a Delta table have statistics collected.
```
Comment (imback82) [inline docs/source/delta-clustering.md:80]:
```suggestion
- You can only specify columns with statistics collected for clustering columns. By default, the first 32 columns in a Delta table have statistics collected.
```

File: docs/source/delta-clustering.md [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+- You can specify up to 4 columns as clustering keys.
```
Comment (imback82) [inline docs/source/delta-clustering.md:81]:
```suggestion
- You can specify up to 4 clustering columns.
```

**Review by imback82 — APPROVED** [review]

**Review by vkorukanti — APPROVED** [review]

**Pull request merged**